Created Online Dictionary App

mikhailDyoshin opened the pull request:
The app is used to search definitions of english words.

Commit:
Initial commit

Created OnlineDictionaryAndroidAppProject and updated .gitignore file.

Commit:
Created Online Dictionary App, which is ready to use.

dmisuvorov commented on app/src/main/java/com/example/onlinedictionaryandroidappproject/data/repository/DictionaryRepositoryImpl.kt:
```diff
@@ -0,0 +1,103 @@
+        )
+    }
+
+    private fun getMeanings(word: WordStorageModel): MutableList<MeaningsDomainModel> {
```
Кажется, что здесь спокойно можно возвращать List вместо MutableList как и в методе getDefinitions, и в моделе безопаснее хранить неизменяемый список

dmisuvorov replied:
Не особо нравится что на data слое оперируем моделью domain уровня (причем не нашел ее (MeaningsDomainModel) использвания в domain)

dmisuvorov commented on app/src/main/java/com/example/onlinedictionaryandroidappproject/presentation/WordFragment.kt:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    private fun getDefinitions(words: List<WordState>): List<String> {
+
```
Бизнес логика в UI слое, надо перенести во ViewModel

Commit:
Started to implement navigation and dynamic lists.

mikhailDyoshin committed:
Continue to implement navigation feature and dinamic lists.

Configured 3 of 4 fragments of the app.

Commit:
Finished implementing navigation feature.

It's needed to add styles to the app.

Commit:
Implemented navigation feature in the app.

Now the app has 4 screens that display different data about requested word such as: part of speach, definitions, examples, synonyms and antonyms.

mikhailDyoshin commented:
My comment)

Commit:
Fixed some issues.

Changed mutable types in models to immutable ones; moved logic, that pulls data from the viewModel to fragments, from fragments to the viewModel. Updated rendering of antonyms and synonyms lists in definition-detail fragment.

dmisuvorov commented on app/src/main/java/com/example/onlinedictionaryandroidappproject/presentation/adapter/DefinitionsListAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+
+class DefinitionsListAdapter(private val dataSet: List<String>) :
+    RecyclerView.Adapter<DefinitionsListAdapter.ViewHolder>() {
+
```
Почитай про ListAdapter, чаще всего лучше использовать его
https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter

dmisuvorov commented on app/src/main/java/com/example/onlinedictionaryandroidappproject/presentation/adapter/WordsListAdapter.kt:
Как совет: после изменений каких то в файле делай Code->Reformat code (Ctrl+Alt+L)

dmisuvorov commented on app/src/main/java/com/example/onlinedictionaryandroidappproject/presentation/fragment/GetWordFragment.kt:
```diff
@@ -0,0 +1,69 @@
+        viewModel.wordState.observe(viewLifecycleOwner) { newState ->
+
+            val adapter =
+                WordsListAdapter(viewModel.getWords(newState))
```
Вот здесь code smell: в классическом MVVM view (в данном случае фрагмент) связывается с ViewModel через некий механизм связывания (LiveData/flow) но напрямую использовать геттеры из ViewModel это больше похоже на MVP
Я бы предложил убрать из конструктора адаптера массив (переделать адаптер как я писал в ListAdapter) и вызывать submtiList в методе observe

mikhailDyoshin replied:
Правильно понял, что вместо геттеров нужно обращаться к newState в лямбда-функции и передавать полученные данные в submitList?

dmisuvorov replied:
Ага все верно

dmisuvorov commented on app/src/main/java/com/example/onlinedictionaryandroidappproject/presentation/viewmodel/WordViewModel.kt:
```diff
@@ -0,0 +1,187 @@
+                        data = RequestState(
+                            statusMessage = "",
+                            wordsStatesList = transformResponse(result.data!!)
+                        )
```
Стараемся убрать !!

mikhailDyoshin replied:
Не понял, что именно убрать)

dmisuvorov replied:
оператор !! Лучше сделай доп проверку на null и если null показ алерта с ошибкой

mikhailDyoshin replied:
Aaa)))

Commit:
Fixed some issues

-discarded code smell in fragments; -replaced RecyclerView.Adapter with ListAdapter; -replaced for-cycles in functions with map-function;

Commit:
Started to implement phonetics-audio-playback-feature.

Commit:
Implemented phonetics-audio-playback-feature.

Now in the Word Detail Fragment a user can learn how an english word is pronounced in few different countries, such as UK, USA, Canada and Australia.

dmisuvorov commented on app/src/main/java/com/example/onlinedictionaryandroidappproject/presentation/viewmodel/WordViewModel.kt:
```diff
@@ -0,0 +1,165 @@
+    private fun transformAudios(word: WordDomainModel): List<AudioState> {
+        val phoneticAudios = word.phoneticAudios
+
+        return phoneticAudios.filter { it.audioURL != null && it.audioURL!!.isNotEmpty() }
```
Здесь можно сократить !it.audioURL.isNullOrEmpty()

Commit:
Refactored some code

Commit:
Started to apply a design.

Commit:
Applied styles to the app header, the loading status,the  message status, the edit-text and the search-button widgets.

Commit:
Apllied styles to list items in WordsListAdapter.

Commit:
Removed top navigation bar, added to the bottom of the screen a home-navigation-button.

Commit:
Applied new styles to definition-items

Commit:
Changed words rendering

Now if the only one word is found there will be no number before it in a list item

Commit:
Finished to apply styles.

The last updates are: applied styles to headers, dividers and list-items in the definition-detail fragment. Changed synonyms and antonyms rendeing logic. Added margit-top to the phonetics-header in the word-detail fragment.

Commit:
Changed colors of the status and the navigation bars.
